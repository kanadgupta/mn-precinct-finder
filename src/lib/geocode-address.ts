import { Client, ReverseGeocodingLocationType } from '@googlemaps/google-maps-services-js';
import { getRuntimeKey } from 'hono/adapter';

import buildMapsUrl from './build-maps-url.js';
import GeocodingError from './errors.js';
import findPrecinct, { type ExtendedPrecinctProps } from './find-precinct.js';
import shortenAddress from './shorten-address.js';

const googlemaps = new Client({});

interface GeocodeResult {
  address: string;
  cache: 'HIT' | 'MISS';
  gmaps: string;
  precinct: ExtendedPrecinctProps;
  type: 'success';
}

type GoogleMapsEnv = Pick<Env, 'GOOGLE_MAPS_API_KEY' | 'GOOGLE_MAPS_RESULTS'>;

/**
 * Takes an address string and returns a formatted address, a Google Maps URL, and the corresponding precinct.
 */
export async function forwardGeocode(
  /** The current environment. May or may not contain Cloudflare KV bindings */
  env: GoogleMapsEnv,
  address: string,
): Promise<GeocodeResult> {
  const runtime = getRuntimeKey();

  // if we're on Cloudflare Workers and have a KV binding, check for a cached result and return it if found
  if (runtime === 'workerd' && env.GOOGLE_MAPS_RESULTS) {
    const cached = await env.GOOGLE_MAPS_RESULTS.get(address);
    if (cached) {
      try {
        return { ...JSON.parse(cached), cache: 'HIT' } as GeocodeResult;
      } catch {
        // delete bad record and fetch a new result
        await env.GOOGLE_MAPS_RESULTS.delete(address);
      }
    }
  }

  return googlemaps
    .geocode({
      adapter: 'fetch',
      params: {
        address,
        bounds: { southwest: { lat: 44.889222, lng: -93.330446 }, northeast: { lat: 45.055223, lng: -93.20494 } },
        key: env.GOOGLE_MAPS_API_KEY,
      },
    })
    .then(async ({ data }) => {
      if (data?.status !== 'OK') throw new GeocodingError(data.results || [], address);

      const { results } = data;

      const filtered = results.filter(result => result?.geometry?.location_type === 'ROOFTOP');

      if (filtered.length !== 1) throw new GeocodingError(filtered || [], address);

      const {
        geometry: { location },
        formatted_address: formattedAddress,
        place_id: placeId,
      } = filtered[0];

      try {
        const precinct = findPrecinct([location.lng, location.lat]);
        const gmaps = buildMapsUrl(formattedAddress, placeId);
        const result = {
          address: shortenAddress(formattedAddress),
          cache: 'MISS',
          gmaps,
          precinct,
          type: 'success',
        } as const satisfies GeocodeResult;

        // cache the result if we're on Cloudflare Workers and have a KV binding
        if (runtime === 'workerd' && env.GOOGLE_MAPS_RESULTS) {
          await env.GOOGLE_MAPS_RESULTS.put(address, JSON.stringify(result));
        }
        return result;
      } catch {
        throw new GeocodingError([], address);
      }
    })
    .catch(err => {
      if (err instanceof GeocodingError) throw err;
      throw new GeocodingError([], address, {
        message: `Our geocoder ran into an unexpected issue (${err.message})`,
        status: 500,
      });
    });
}

/**
 * Takes latitude and longitude coordinates and returns a formatted address and a Google Maps URL.
 *
 * @deprecated I don't think we need to support this anymore.
 */
export async function reverseGeocode(
  /** The current environment. May or may not contain Cloudflare KV bindings */
  env: GoogleMapsEnv,
  long: number | string,
  lat: number | string,
) {
  return googlemaps
    .reverseGeocode({
      params: {
        key: env.GOOGLE_MAPS_API_KEY,
        latlng: `${lat}, ${long}`,
        location_type: [ReverseGeocodingLocationType.ROOFTOP],
      },
    })
    .then(({ data }) => {
      const { results } = data;

      const { formatted_address: formattedAddress, place_id: placeId } = results[0];
      const gmaps = buildMapsUrl(formattedAddress, placeId);

      return { address: shortenAddress(formattedAddress), gmaps };
    });
}
