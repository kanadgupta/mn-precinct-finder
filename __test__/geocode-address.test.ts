import nock from 'nock';
import { afterAll, beforeAll, describe, expect, it } from 'vitest';

import { forwardGeocode, reverseGeocode } from '../src/lib/geocode-address.js';

describe('#forwardGeocode', () => {
  beforeAll(() => {
    nock.disableNetConnect();
  });

  afterAll(() => {
    nock.enableNetConnect();
  });

  it('should return valid data for valid query', async () => {
    nock('https://maps.googleapis.com:443')
      .get('/maps/api/geocode/json')
      .query(true)
      .reply(200, {
        results: [
          {
            formatted_address: 'some formatted address',
            geometry: {
              location: { lat: 44.9778, lng: -93.265 },
              location_type: 'ROOFTOP',
            },
            place_id: 'test',
          },
        ],
        status: 'OK',
      });

    await expect(forwardGeocode({ GOOGLE_MAPS_API_KEY: '' }, 'unformatted address')).resolves.toMatchSnapshot();
  });

  it('should throw if response returns empty results', async () => {
    nock('https://maps.googleapis.com:443').get('/maps/api/geocode/json').query(true).reply(200, {
      results: [],
      status: 'ZERO_RESULTS',
    });

    await expect(forwardGeocode({ GOOGLE_MAPS_API_KEY: '' }, 'zero results address')).rejects.toMatchInlineSnapshot(
      "[GeocodingError: We weren't able to find any addresses in Minnesota for 'zero results address'.]",
    );
  });

  it('should throw if response is unknown error', async () => {
    nock('https://maps.googleapis.com:443').get('/maps/api/geocode/json').query(true).reply(400, {
      error_message: 'some error message',
      status: 'UNKNOWN_ERROR',
    });

    await expect(forwardGeocode({ GOOGLE_MAPS_API_KEY: '' }, 'unknown error address')).rejects.toMatchInlineSnapshot(
      '[GeocodingError: Our geocoder ran into an unexpected issue (Request failed with status code 400)]',
    );
  });

  it('should throw if none of the results are of the rooftop location type', async () => {
    nock('https://maps.googleapis.com:443')
      .get('/maps/api/geocode/json')
      .query(true)
      .reply(200, {
        results: [
          {
            formatted_address: 'some formatted address',
            geometry: {
              location: { lat: 44.9778, lng: -93.265 },
              location_type: 'RANGE_INTERPOLATED',
            },
            place_id: 'test',
          },
        ],
        status: 'OK',
      });

    await expect(forwardGeocode({ GOOGLE_MAPS_API_KEY: '' }, 'non-rooftop address')).rejects.toMatchInlineSnapshot(
      "[GeocodingError: We weren't able to find any addresses in Minnesota for 'non-rooftop address'.]",
    );
  });

  it('should throw if geocoded result does not yield a precinct (i.e. outside of MN)', async () => {
    nock('https://maps.googleapis.com:443')
      .get('/maps/api/geocode/json')
      .query(true)
      .reply(200, {
        results: [
          {
            formatted_address: 'some formatted address',
            geometry: {
              location: { lat: 0, lng: 0 },
              location_type: 'ROOFTOP',
            },
            place_id: 'test',
          },
        ],
        status: 'OK',
      });

    await expect(forwardGeocode({ GOOGLE_MAPS_API_KEY: '' }, 'non-MN address')).rejects.toMatchInlineSnapshot(
      "[GeocodingError: We weren't able to find any addresses in Minnesota for 'non-MN address'.]",
    );
  });

  it('should throw if multiple valid results', async () => {
    const results = [
      {
        formatted_address: 'some formatted address',
        geometry: {
          location: { lat: 44.9778, lng: -93.265 },
          location_type: 'ROOFTOP',
        },
        place_id: 'test',
      },
      {
        formatted_address: 'another formatted address',
        geometry: {
          location: { lat: 44.9778, lng: -93.265 },
          location_type: 'ROOFTOP',
        },
        place_id: 'test-again',
      },
    ];

    nock('https://maps.googleapis.com:443').get('/maps/api/geocode/json').query(true).reply(200, {
      results,
      status: 'OK',
    });

    await expect(forwardGeocode({ GOOGLE_MAPS_API_KEY: '' }, 'non-MN address')).rejects.toMatchInlineSnapshot(
      "[GeocodingError: We found multiple matches for 'non-MN address'.]",
    );
  });
});

describe('#reverseGeocode', () => {
  it('should properly reverse geocode result', async () => {
    nock('https://maps.googleapis.com:443')
      .get('/maps/api/geocode/json')
      .query(true)
      .reply(200, {
        results: [
          {
            formatted_address: 'a reverse geocoded address',
            geometry: {
              location: { lat: 0, lng: 0 },
              location_type: 'ROOFTOP',
            },
            place_id: 'test',
          },
        ],
        status: 'OK',
      });

    await expect(reverseGeocode({ GOOGLE_MAPS_API_KEY: '' }, 0, 0)).resolves.toMatchSnapshot();
  });
});
