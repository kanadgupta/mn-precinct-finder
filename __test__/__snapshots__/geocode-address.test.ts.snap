// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#forwardGeocode > should return valid data for valid query 1`] = `
{
  "address": "some formatted address",
  "cache": "MISS",
  "gmaps": "https://www.google.com/maps/search/?api=1&query=some+formatted+address&query_place_id=test",
  "precinct": {
    "CongDist": "5",
    "County": "Hennepin",
    "CountyID": "27",
    "CtyComDist": "03",
    "Hospital": "No data",
    "Judicial": "04",
    "MCDCode": "135",
    "MCDName": "Minneapolis",
    "MNLegDist": "59B",
    "MNSenDist": "59",
    "Park": "4",
    "Precinct": "Minneapolis W-7 P-11",
    "PrecinctCode": "1705",
    "PrecinctID": "270531705",
    "SoilAndWater": "No data",
    "Ward": "W-07",
  },
  "type": "success",
}
`;

exports[`#reverseGeocode > should properly reverse geocode result 1`] = `
{
  "address": "a reverse geocoded address",
  "gmaps": "https://www.google.com/maps/search/?api=1&query=a+reverse+geocoded+address&query_place_id=test",
}
`;
